File: minimumRemoveToMakeValidParentheses.go
// problem: https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/
// Runtime: 8 ms, faster than 88.19% of Go online submissions for Minimum Remove to Make Valid Parentheses.
// Memory Usage: 7.1 MB, less than 40.97% of Go online submissions for Minimum Remove to Make Valid Parentheses.

package algorithms

func minRemoveToMakeValid(s string) string {
	chars := []rune{}
	for _, c := range s {
		chars = append(chars, c)
	}
	stack := []int{}
	for i, c := range chars {
		if c == '(' {
			stack = append(stack, i)
		} else if len(stack) > 0 && c == ')' {
			stack = stack[:len(stack)-1]
		} else if c == ')' {
			chars[i] = '*'
		}
	}
	for _, j := range stack {
		chars[j] = '*'
	}
	answer := []rune{}
	for _, c := range chars {
		if c != '*' {
			answer = append(answer, c)
		}
	}
	return string(answer)
}

File: minimumRemoveToMakeValidParentheses.py
# problem: https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/
# Runtime: 84 ms, faster than 94.25% of Python3 online submissions for Minimum Remove to Make Valid Parentheses.
# Memory Usage: 16 MB, less than 68.16% of Python3 online submissions for Minimum Remove to Make Valid Parentheses.


class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        chars = list(s)
        stack = []
        for i, c in enumerate(chars):
            if c == "(":
                stack.append(i)
            elif c == ")" and stack:
                stack.pop()
            elif c == ")":
                chars[i] = ""
        for i in stack:
            chars[i] = ""
        return "".join(chars)
